File: test-connectivity.js
#!/usr/bin/env node

const net = require('net');
const { MongoClient } = require('mongodb');
const amqp = require('amqplib');
const Minio = require('minio');

// Test configuration
const tests = {
  mongodb: {
    host: 'localhost',
    port: 27017,
    uri: 'mongodb://admin:password@localhost:27017/helmet_detection?authSource=admin'
  },
  rabbitmq: {
    host: 'localhost',
    port: 5672,
    url: 'amqp://guest:guest@localhost:5672'
  },
  minio: {
    host: 'localhost',
    port: 9000,
    endpoint: 'localhost',
    accessKey: 'minioadmin',
    secretKey: 'minioadmin'
  }
};

// Test port connectivity
function testPort(host, port, name) {
  return new Promise((resolve, reject) => {
    const socket = new net.Socket();
    
    socket.setTimeout(5000);
    
    socket.on('connect', () => {
      console.log(`✅ ${name} port ${port} is reachable`);
      socket.destroy();
      resolve();
    });
    
    socket.on('timeout', () => {
      console.log(`❌ ${name} port ${port} timeout`);
      socket.destroy();
      reject(new Error(`${name} timeout`));
    });
    
    socket.on('error', (err) => {
      console.log(`❌ ${name} port ${port} error: ${err.message}`);
      reject(err);
    });
    
    socket.connect(port, host);
  });
}

// Test MongoDB connection
async function testMongoDB() {
  try {
    const client = new MongoClient(tests.mongodb.uri);
    await client.connect();
    await client.db('helmet_detection').admin().ping();
    console.log('✅ MongoDB connection successful');
    await client.close();
  } catch (error) {
    console.log(`❌ MongoDB connection failed: ${error.message}`);
    throw error;
  }
}

// Test RabbitMQ connection
async function testRabbitMQ() {
  try {
    const connection = await amqp.connect(tests.rabbitmq.url);
    const channel = await connection.createChannel();
    console.log('✅ RabbitMQ connection successful');
    await channel.close();
    await connection.close();
  } catch (error) {
    console.log(`❌ RabbitMQ connection failed: ${error.message}`);
    throw error;
  }
}

// Test MinIO connection
async function testMinIO() {
  try {
    const minioClient = new Minio.Client({
      endPoint: tests.minio.endpoint,
      port: tests.minio.port,
      useSSL: false,
      accessKey: tests.minio.accessKey,
      secretKey: tests.minio.secretKey
    });
    
    // Test bucket listing
    const buckets = await minioClient.listBuckets();
    console.log('✅ MinIO connection successful');
    console.log(`   Found ${buckets.length} buckets`);
  } catch (error) {
    console.log(`❌ MinIO connection failed: ${error.message}`);
    throw error;
  }
}

// Run all tests
async function runTests() {
  console.log('🔍 Testing connectivity to Docker Compose services...\n');
  
  try {
    // Test basic port connectivity
    console.log('📡 Testing port connectivity...');
    await testPort(tests.mongodb.host, tests.mongodb.port, 'MongoDB');
    await testPort(tests.rabbitmq.host, tests.rabbitmq.port, 'RabbitMQ');
    await testPort(tests.minio.host, tests.minio.port, 'MinIO');
    
    console.log('\n🔌 Testing application connectivity...');
    // Test application-level connectivity
    await testMongoDB();
    await testRabbitMQ();
    await testMinIO();
    
    console.log('\n🎉 All connectivity tests passed!');
    console.log('\nYou can now run your backend and ai-service locally:');
    console.log('  Backend: cd backend && npm run dev');
    console.log('  AI Service: cd ai-service && python main.py');
    
  } catch (error) {
    console.log('\n❌ Some tests failed. Please check that Docker Compose is running:');
    console.log('  docker-compose -f docker-compose.dev.yml up -d');
    process.exit(1);
  }
}

runTests(); 

File: test-connectivity.py
#!/usr/bin/env python3

import socket
import sys
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv('local.env')

def test_port_connectivity(host, port, service_name):
    """Test if a port is reachable"""
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)
        result = sock.connect_ex((host, port))
        sock.close()
        
        if result == 0:
            print(f"✅ {service_name} port {port} is reachable")
            return True
        else:
            print(f"❌ {service_name} port {port} is not reachable")
            return False
    except Exception as e:
        print(f"❌ {service_name} connection error: {e}")
        return False

def test_rabbitmq_connection():
    """Test RabbitMQ connection"""
    try:
        import pika
        
        rabbitmq_host = os.getenv('RABBITMQ_HOST', 'localhost')
        rabbitmq_port = int(os.getenv('RABBITMQ_PORT', '5672'))
        rabbitmq_user = os.getenv('RABBITMQ_USER', 'guest')
        rabbitmq_password = os.getenv('RABBITMQ_PASSWORD', 'guest')
        
        credentials = pika.PlainCredentials(rabbitmq_user, rabbitmq_password)
        parameters = pika.ConnectionParameters(
            host=rabbitmq_host,
            port=rabbitmq_port,
            credentials=credentials
        )
        
        connection = pika.BlockingConnection(parameters)
        channel = connection.channel()
        
        print("✅ RabbitMQ connection successful")
        connection.close()
        return True
        
    except ImportError:
        print("⚠️  pika not installed, installing...")
        os.system("pip install pika")
        return test_rabbitmq_connection()
    except Exception as e:
        print(f"❌ RabbitMQ connection failed: {e}")
        return False

def test_minio_connection():
    """Test MinIO connection"""
    try:
        from minio import Minio
        
        minio_endpoint = os.getenv('MINIO_ENDPOINT', 'localhost:9000')
        minio_access_key = os.getenv('MINIO_ACCESS_KEY', 'minioadmin')
        minio_secret_key = os.getenv('MINIO_SECRET_KEY', 'minioadmin')
        
        client = Minio(
            minio_endpoint,
            access_key=minio_access_key,
            secret_key=minio_secret_key,
            secure=False
        )
        
        # Test connection by listing buckets
        buckets = list(client.list_buckets())
        print("✅ MinIO connection successful")
        print(f"   Found {len(buckets)} buckets")
        return True
        
    except ImportError:
        print("⚠️  minio not installed, installing...")
        os.system("pip install minio")
        return test_minio_connection()
    except Exception as e:
        print(f"❌ MinIO connection failed: {e}")
        return False

def main():
    """Run all connectivity tests"""
    print("🔍 Testing AI Service connectivity to Docker Compose services...\n")
    
    # Load configuration from environment
    rabbitmq_host = os.getenv('RABBITMQ_HOST', 'localhost')
    rabbitmq_port = int(os.getenv('RABBITMQ_PORT', '5672'))
    minio_endpoint = os.getenv('MINIO_ENDPOINT', 'localhost:9000')
    minio_host = minio_endpoint.split(':')[0]
    minio_port = int(minio_endpoint.split(':')[1])
    
    all_tests_passed = True
    
    # Test port connectivity
    print("📡 Testing port connectivity...")
    tests = [
        (rabbitmq_host, rabbitmq_port, "RabbitMQ"),
        (minio_host, minio_port, "MinIO")
    ]
    
    for host, port, service in tests:
        if not test_port_connectivity(host, port, service):
            all_tests_passed = False
    
    # Test application connectivity
    print("\n🔌 Testing application connectivity...")
    
    if not test_rabbitmq_connection():
        all_tests_passed = False
    
    if not test_minio_connection():
        all_tests_passed = False
    
    print("\n" + "="*50)
    if all_tests_passed:
        print("🎉 All AI Service connectivity tests passed!")
        print("\nYou can now run the AI Service locally:")
        print("  cd ai-service && python main.py")
        print("\nOr debug it in VS Code using the 'Debug AI Service' configuration")
    else:
        print("❌ Some tests failed. Please check that Docker Compose is running:")
        print("  docker-compose -f docker-compose.dev.yml up -d")
        sys.exit(1)

if __name__ == "__main__":
    main() 